File: JavaScript/LowestCommonAncestorOfABinarySearchTree/LowestCommonAncestorOfABinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * (202ms)
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    while((root.val - p.val) * (root.val - q.val) > 0) {
        root = p.val < root.val ? root.left : root.right;
    }
    return root;
};

File: Python/LowestCommonAncestorOfABinarySearchTree/LowestCommonAncestorOfABinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
	# (125ms) iteratively solution
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        while (root.val - p.val) * (root.val - q.val) > 0:
            root = root.left if p.val < root.val else root.right
        return root
		
	# (129ms)
	def lowestCommonAncestor2(self, root, p, q):
		return root if (root.val - p.val) * (root.val - q.val) <= 0 else \
		    self.lowestCommonAncestor(root.left if p.val < root.val else root.right, p, q)